providers/duo/v4.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return Session{}, err
+	}
+
+	form, ok = findFirstForm(body)
+	if !ok {
+		return Session{}, ErrCouldNotFindDuoForm
+	}
+
+	session := Session{
+		Sid:  form.Inputs["sid"],
+		Xsrf: form.Inputs["_xsrf"],
+	}
+
+	return session, nil
+}
+
+func (d DuoV4) GetRegisteredFactorDataRequest(ctx context.Context, session Session) (*http.Request, error) {
+	vals := url.Values{
+		"sid":              []string{session.Sid},
+		"post_auth_action": []string{"OIDC_EXIT"},
+	}
+
+	next := d.BaseURL.ResolveReference(&url.URL{Path: "/frame/v4/auth/prompt/data"})
+	next.RawQuery = vals.Encode()
+	return http.NewRequestWithContext(ctx, "GET", next.String(), nil)
+}
+
+func (d DuoV4) GetRegisteredFactors(ctx context.Context, session Session) ([]Factor, error) {
+	dataReq, _ := d.GetRegisteredFactorDataRequest(ctx, session)
+	resp, err := d.Client.Do(dataReq)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	var data DuoV4APIResponse[RegisteredFactors]
+	return data.Response.Phones, json.NewDecoder(resp.Body).Decode(&data)
+}
+
+// Exit ends the current OIDC session. The returned http.Response will contain the response of whereever Duo redirected us, usually to the Redirect URI of the OIDC application.
+func (d DuoV4) Exit(ctx context.Context, session Session, deviceKey string) (*http.Response, error) {
+	v := url.Values{
+		"sid":           []string{session.Sid},
+		"txid":          []string{session.Txid},
+		"_xsrf":         []string{session.Xsrf},
+		"device_key":    []string{deviceKey},
+		"factor":        []string{"Duo Push"},
+		"dampen_choice": []string{"false"},
+	}
+
+	next := d.BaseURL.ResolveReference(&url.URL{Path: "/frame/v4/oidc/exit"})
+	req, _ := http.NewRequestWithContext(ctx, "POST", next.String(), strings.NewReader(v.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	return d.Client.Do(req)
+}
+
+// Prompt issues a prompt for the given session to the given factor and device.
+//
+// A new Session is returned, which must be used for future calls.
+func (d DuoV4) Prompt(ctx context.Context, session Session, factor, device string) (Session, error) {
+	vals := url.Values{
+		"sid":    []string{session.Sid},
+		"_xsrf":  []string{session.Xsrf},
+		"factor": []string{factor},
+		"device": []string{device},
+	}
+
+	next := d.BaseURL.ResolveReference(&url.URL{Path: "/frame/v4/prompt"})
+	req, _ := http.NewRequestWithContext(ctx, "POST", next.String(), strings.NewReader(vals.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+	resp, err := d.Client.Do(req)
+	if err != nil {
+		return Session{}, err
+	}
+
+	defer resp.Body.Close()
+	blob, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return Session{}, err
+	}
+
+	session.Txid = gjson.GetBytes(blob, "response.txid").Str
+	return session, nil
+}
+
+func (d DuoV4) PromptPhone1(ctx context.Context, session Session) (Session, error) {
```

Comment:
I know that `phone1` is a test. Can you add a comment to clarify the purpose of this function for new readers?

Reply:
`phone1` is actually not a test :) that is the default name of a device in Duo when using push notifications